python_modules/dagster/dagster/core/asset_defs/decorators.py

```diff
@@ -198,9 +204,29 @@ def __call__(self, fn: Callable) -> AssetsDefinition:
         with warnings.catch_warnings():
             warnings.simplefilter("ignore", category=ExperimentalWarning)
 
+            if check.is_mapping_param(self.io_manager):
+                io_manager = cast(Mapping[str, IOManagerDefinition], self.io_manager)
+                check.invariant(
+                    len(self.io_manager) == 1,
+                    "Attempted to provide multiple io manager keys to asset.",
+                )
+                io_manager_key = list(io_manager.keys())[0]
+                self.resource_defs[io_manager_key] = io_manager[io_manager_key]
+            elif isinstance(self.io_manager, str):
+                io_manager_key = cast(str, self.io_manager)
+            elif self.io_manager is not None:
+                io_manager_def = check.inst_param(
+                    self.io_manager, "io_manager", IOManagerDefinition
+                )
+                out_asset_resource_key = "_".join(out_asset_key.path)
```

Comment:
This is both the main piece and the gnarliest piece. In order to bypass resource uniqueness per key, we need to essentially construct a unique key per asset. I think it extremely unlikely to run into any collisions, but using underscores feels weird. An alternative would be to add an exception to resource key constraint to allow the . syntax for assets. Another solution would be to just paper over this in dagit so that it wasn't exposed to the user as this key, but instead "io manager" or something.

Reply:
Agreed that this is gnarly, though it might still be fine. 

Two questions:
* What if sidestepped resource keys entirely and instead somehow fetched the definition off of the `io_manager` property directly?
* How "reversible" is this approach? If we want to implement this in a different way later, will that cause problems?

I think a small improvement would be to use double-underscores, because that's what we do when auto-constructing the op name and users are unlikely to have double-underscores in their asset names.

Reply:
1. Should be possible to fetch directly off the definition. I think we would add a private accessor or something to the output definition so that it was mostly hidden from end users, and just pull it directly when building the resource requirements. 

2. Hmm since this is all internal implementation details / the user should never be really interacting with the key, I don't think it matters too much. The two cases I can think of this mattering:
- re-execution from a config blob that somehow makes use of the system provided resource key. If we try to re-execute from a config blob constructed when this key existed, and then changed implementation, that would no longer work. 
- Snapshot reading with a new version of dagit and old version of user code server. If we decided to go with an approach that had some sort of special treatment of the io manager in the snapshot, we would have to make sure to handle backcompat there I guess.